kauri/tree.go

```diff
@@ -0,0 +1,170 @@
+package kauri
+
+import (
+	"math"
+
+	"github.com/relab/hotstuff"
+)
+
+// MaxChild determines the branching factor of the tree.
+const MaxChild = 2
+
+// TreeConfiguration is an abstraction for a tree communication model.
+type TreeConfiguration interface {
+	InitializeWithPIDs(ids map[hotstuff.ID]int)
+	GetHeight() int
+	GetChildren() []hotstuff.ID
+	GetSubTreeNodes() []hotstuff.ID
+	GetParent() (hotstuff.ID, bool)
+}
+
+// FaultFreeTree implements a fault free tree configuration.
+type FaultFreeTree struct {
+	ID                  hotstuff.ID
+	ConfigurationLength int
+	height              int
+	idToPosMapping      map[hotstuff.ID]int
+	posToIDMapping      map[int]hotstuff.ID
+}
+
+// CreateTree Creates the tree configuration, currently only fault free tree configuration is supported.
+func CreateTree(configurationLength int, myID hotstuff.ID) TreeConfiguration {
```

Comment:
This is the interface anti-pattern in Go that I mentioned a while back. Generally, we avoid returning an interface type and defining interfaces before they are needed.

Suggest to rename this function to `newFaultFreeTree` and return `*faultFreeTree` instead of the interface. If you later add a new tree type, you can create another construction function that returns the other tree type.